job-runner/chain/traverser.go

```diff
@@ -377,3 +400,47 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
 		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
 	}
 }
+
+// retrySequence retries a sequence by doing the following
+//   1. Mark the failed job and all jobs before it in the sequence to PENDING
+//   state, so that they are ran again on sequence retry. This is OK to do
+//   since all jobs are idempotent.
+//   2. Decrement the number of jobs that were previously ran in the sequence
+//   from the chain job count.
+//   3. Increment retry count for the sequence
+//   4. Enqueue the first job in the sequence to be ran again. It'll
+//   subsequently enqueue the remaining jobs in the sequence.
+func (t *traverser) retrySequence(failedJob proto.Job) error {
+	sequenceStartJob := t.chain.SequenceStartJob(failedJob)
+
+	// Set the state of the failed job in the sequence to "Pending" and decrement
+	// number of jobs ran from chain.
+	t.chain.SetJobState(failedJob.Id, proto.STATE_PENDING)
+	t.chain.n -= 1
+
+	// For the rest of the jobs in the sequence that were previously completed,
+	// set the state of the job to "Pending" and decrement number of jobs ran
+	// from chain
+	currJob := sequenceStartJob
+	for currJob.Id != failedJob.Id {
+		// Mark failed job as pending and update chain job count
+		t.chain.SetJobState(currJob.Id, proto.STATE_PENDING)
+		t.chain.n -= 1
+
+		// Update currJob to next job in sequence
+		nextJobs := t.chain.NextJobs(currJob.Id)
+		if len(nextJobs) != 1 {
+			err := fmt.Errorf("traversing sequence and expected next jobs list to be length of 1 but it is lenght %d", len(nextJobs))
```

Comment:
Is this really possible? This is the only possible error return in this func, but we already have a known-good job chain that's already been traversed so... why are we checking for this?  Because if we not check for errors, then it simplifies the caller a little, too (i.e. my previous comment re logging seq retry before calling this func).

Reply:
we chatted offline. it's possible for the sequence retry to include a nested sequence. removed this length check.